### PR TITLE
Small improvement for String.length

### PR DESCRIPTION
**Pull request by abelbraaksma**
The PR removes some overhead for `String.length`, see https://github.com/dotnet/fsharp/issues/9390#issuecomment-643780818.

The disassembly will change from this:

```assembly
test      rax,rax              ; rax & rax, in other words: null-check, sets flags
je        short M00_L01        ; if zero (null) jump back to loop
cmp       dword ptr [rax+8],0  ; compare string-length to 0
jbe       short M00_L01        ; if equal, jump back to loop
mov       ecx,[rax+8]          ; if not, get string length fallthrough to loop
```

to this:

```assembly
test      rax,rax        ; rax & rax, for null-check, sets flags
je        short M00_L01  ; if zero (null) jump back to loop
mov       ecx,[rax+8]    ; get the string-length from the string and fallthrough to loop
```

Which, in timings, gives roughly a 33% improvement, but since this is hard to test in isolation, it may be more because more chances exist now for optimizing & inlining this by the JIT.

More info and timings in the connected issue comment above.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9469) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9469) before we can accept your contribution.<br/><br/>:x: abelbraaksma [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=9469)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=9469) it.</sub>

**Comment by abelbraaksma** [comment]
Hmm, I've contributed before, though it's a while ago that I made PR's. From time to time you have to sign again?

image

**Comment by cartermp** [comment]
The CLA is a consequence of moving to dotnet/fsharp I believe?

**Comment by abelbraaksma** [comment]
Possible, anyway, signing took 3 seconds, so it's not a big deal :)

File: src/fsharp/FSharp.Core/string.fs [hunk 1]
```diff
@@ -31,12 +31,15 @@ namespace Microsoft.FSharp.Core
 
         [<CompiledName("Map")>]
         let map (mapping: char -> char) (str:string) =
-            if String.IsNullOrEmpty str then
-                String.Empty
+            if String.IsNullOrEmpty str then String.Empty
```
Comment (forki) [inline src/fsharp/FSharp.Core/string.fs:34]: Please keep the original indentation style.

File: src/fsharp/FSharp.Core/string.fs [hunk 2]
```diff
@@ -31,12 +31,15 @@ namespace Microsoft.FSharp.Core
...
-                let res = StringBuilder str.Length
-                str |> iter (fun c -> res.Append(mapping c) |> ignore)
-                res.ToString()
+                let result = str.ToCharArray()
+                let mutable i = 0
+                for c in result do
+                    result.[i] <- mapping c
+                    i <- i + 1
+
+                String result
```
Comment (forki) [inline src/fsharp/FSharp.Core/string.fs:42]: Please use explicit tostring()

File: src/fsharp/FSharp.Core/string.fs [hunk 3]
```diff
@@ -104,7 +107,5 @@ namespace Microsoft.FSharp.Core
 
         [<CompiledName("Length")>]
         let length (str:string) =
-            if String.IsNullOrEmpty str then
-                0
-            else
-                str.Length
+            if isNull str then 0
```
Comment (forki) [inline src/fsharp/FSharp.Core/string.fs:110]: This is unrelated

Comment (forki) [reply src/fsharp/FSharp.Core/string.fs:110]: Well, according to the pr text this is the related change and the stuff in map is unrelated

**Comment by KevinRansom** [comment]
@abelbraaksma It seems as though you combined two changes into one PR and proposed it twice.

/cc @vzarytovskii

**Comment by abelbraaksma** [comment]
@KevinRansom, yeah, that's why I shouldn't do this long after midnight ;)

I meant to split them (@cartermp and @forkiasked for a PR per function), and thought I only locally had done the accidental merge.

I used VS's github integration, and couldn't switch branches on the PR page, so I couldn't create a new PR on the 2nd branch, and after some frustration clearly did something else wrong ;)

Will fix.

File: src/fsharp/FSharp.Core/string.fs (hunk 3 above)
Comment (abelbraaksma) [reply src/fsharp/FSharp.Core/string.fs:110]: @forki Indeed, sorry for the confusion, will fix that (if I manage to get VS to switch between PRs).

File: src/fsharp/FSharp.Core/string.fs [hunk 4]
```diff
@@ -104,7 +104,5 @@ namespace Microsoft.FSharp.Core
...
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/string.fs:107]: The change is good. Typically this would be formatted to a single line in this codebase, though. If you make that change I'll approve this.

Comment (abelbraaksma) [reply src/fsharp/FSharp.Core/string.fs]: @cartermp, thx, I forgot, @forki mentioned that too. Fixed in https://github.com/dotnet/fsharp/pull/9469/commits/91ca1d4891aed625c5e5f1e821da649015f8acab

File: src/fsharp/FSharp.Core/string.fs (hunk 2 above)
Comment (abelbraaksma) [reply src/fsharp/FSharp.Core/string.fs]: @forki, tx, but that would've output `"System.Char[]"`, but I've settled in https://github.com/dotnet/fsharp/pull/9470 with `new String(result)` to make it abundantly clear that the constructor is needed here.

In this PR it is not relevant anymore.

File: src/fsharp/FSharp.Core/string.fs (hunk 1 above)
Comment (abelbraaksma) [reply src/fsharp/FSharp.Core/string.fs]: Resolved in https://github.com/dotnet/fsharp/pull/9470, where it belonged. Sorry for the noise.

**Comment by KevinRansom** [comment]
@abelbraaksma thanks for taking care of this.

Kevin
